pkg/reconciler/workload/placement/placement_reconcile_scheduling.go

```diff
@@ -111,15 +130,92 @@ func (r *placementSchedulingReconciler) getAllValidSyncTargetsForPlacement(clust
+	locationSyncTargets, err := locationreconciler.LocationSyncTargets(syncTargets, location)
+	if err != nil {
+		return locationWorkspace, nil, err
+	}
+	if len(locationSyncTargets) == 0 {
+		conditions.MarkFalse(placement,
+			schedulingv1alpha1.PlacementScheduled,
+			schedulingv1alpha1.ScheduleNoValidTargetReason,
+			conditionsv1alpha1.ConditionSeverityWarning,
+			"No SyncTarget in the selected Location",
+		)
+		return locationWorkspace, nil, nil
+	}
+
+	validSyncTargets, err := r.filterAPICompatible(ctx, clusterName, placement, locationSyncTargets)
 	if err != nil {
 		return locationWorkspace, nil, err
 	}
 
+	if len(validSyncTargets) == 0 {
+		return locationWorkspace, validSyncTargets, nil
+	}
+
 	// find all the valid sync targets.
-	validClusters := locationreconciler.FilterNonEvicting(locationreconciler.FilterReady(locationClusters))
+	validSyncTargets = locationreconciler.FilterNonEvicting(locationreconciler.FilterReady(validSyncTargets))
+	if len(validSyncTargets) == 0 {
+		conditions.MarkFalse(placement,
+			schedulingv1alpha1.PlacementScheduled,
+			schedulingv1alpha1.ScheduleNoValidTargetReason,
+			conditionsv1alpha1.ConditionSeverityWarning,
+			"No SyncTarget is ready or non evicting",
+		)
+	}
+
+	return locationWorkspace, validSyncTargets, nil
+}
+
+func (r *placementSchedulingReconciler) filterAPICompatible(ctx context.Context, clusterName logicalcluster.Name, placement *schedulingv1alpha1.Placement, syncTargets []*workloadv1alpha1.SyncTarget) ([]*workloadv1alpha1.SyncTarget, error) {
+	logger := klog.FromContext(ctx)
+	var filteredSyncTargets []*workloadv1alpha1.SyncTarget
+
+	apiBindings, err := r.listWorkloadAPIBindings(clusterName)
+	if err != nil {
+		return filteredSyncTargets, err
+	}
+
+	var messages []string
+	for _, syncTargert := range syncTargets {
+		supportedAPIMap := map[apisv1alpha1.GroupResource]workloadv1alpha1.ResourceToSync{}
+		for _, resource := range syncTargert.Status.SyncedResources {
+			if resource.State == workloadv1alpha1.ResourceSchemaAcceptedState {
+				supportedAPIMap[resource.GroupResource] = resource
+			}
+		}
+
+		supported := true
+		for _, binding := range apiBindings {
+			for _, desiredAPI := range binding.Status.BoundResources {
+				supportedAPI, ok := supportedAPIMap[apisv1alpha1.GroupResource{
+					Group:    desiredAPI.Group,
+					Resource: desiredAPI.Resource,
+				}]
+				if !ok || supportedAPI.IdentityHash != desiredAPI.Schema.IdentityHash {
+					supported = false
+					messages = append(messages, fmt.Sprintf("SyncTarget %s does not support APIBinding %s", syncTargert.Name, binding.Name))
+					logger.V(4).Info("Does not support APIBindings", "workspace", clusterName, "APIBinding", binding.Name, "syncTarget", syncTargert.Name)
+					break
+				}
+			}
+		}
+
+		if supported {
+			filteredSyncTargets = append(filteredSyncTargets, syncTargert)
+		}
+	}
+
+	if len(filteredSyncTargets) == 0 {
+		conditions.MarkFalse(placement,
```

Comment:
another one 🔴 . Don't mutate anything when it is not clearly expected by the function name.